iceoryx_utils/include/iceoryx_utils/cxx/forward_list.hpp

```diff
@@ -0,0 +1,382 @@
+        friend class const_iterator;
+        forward_list<T, CAPACITY>* m_list;
+        size_type m_iterListNodeIdx;
+
+    }; // class iterator
+
+    /// @brief nested const_iterator class, --> linked data element is 'const'
+    class const_iterator
+    {
+      public:
+        // provide the following public types for a std::iterator compatible iterator_category interface
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = const T;
+        using difference_type = void;
+        using pointer = const T*;
+        using reference = const T&;
+
+      private:
+        /// @brief private construct for an iterator, the iterator is bundled to
+        ///         an existing parent (object) of type forward_list,
+        ///         an iterator is only constructed through calls to before_begin(), begin() or end()
+        /// @param[in] parent is the const forward_list the this iterator operates on
+        /// @param[in] idx is the index of the list element (within allocated memory of parent list)
+        explicit const_iterator(const forward_list* parent, size_type idx) noexcept;
+
+      public:
+        /// @brief construct a const_iterator from an (non-const_) iterator
+        /// @param[in] iter is the iterator which will deliver list and index info for the const_iterator
+        const_iterator(const iterator iter) noexcept;
+
+
+        /// @brief prefix increment iterator, so it points to the next list element
+        ///         when trying to increment beyond the end of the list,
+        ///         iterator stays pointing at the end and a message is forwarded to the error_message
+        ///         handler / cerr stream
+        /// @return reference to this iterator object
+        const_iterator& operator++() noexcept;
+
+
+        /// @brief comparing list iterators on equality
+        ///         the referenced list position is compared, not the content of the list element (T-typed)
+        ///         there is no content for meta-elements like before_begin() and end()
+        ///         only iterators of the same parent list can be compared; in case of misuse, terminate() is invoked
+        /// @param[in] rhs_citer is the 2nd iterator to compare to
+        /// @return list position for two iterators is the same (true) or different (false)
+        bool operator==(const const_iterator rhs_citer) const noexcept;
+
+        /// @brief dereferencing element content via iterator-position element
+        /// @return reference to list element data
+        const T& operator*() const noexcept;
+
+        /// @brief dereferencing element content via iterator-position element
+        /// @return pointer to const list data element
+        const T* operator->() const noexcept;
+
+      private:
+        using CListNodePointer = const ListNode*;
+        friend class forward_list<T, CAPACITY>;
+
+        const forward_list<T, CAPACITY>* m_cList;
```

Comment:
I do not think this should be const. It is internal to the iterator and in principle it does not change (but the same is true for iterator, were it is not const), but later on there is a const_cast required due to this. I will mark the occurrence of the const_cast.